### ISSUE TITLE
Adding alternative ordination and clustering methods

### ISSUE THREAD

**Issue by mortonjt** [biocore/emperor#356]
Looking at Visumap, the majority of those algorithms are found in scikit-learn or scikit-bio, namely

__Ordination__
* [SMACOF](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/manifold/mds.py)
* Multidimensional Scaling (Sammon Map)
* [Correspondence Analysis](https://github.com/biocore/scikit-bio/blob/master/skbio/stats/ordination/_canonical_correspondence_analysis.py)

__Classification__
* Linear Discriminate Analysis
* [Nearest Neighbor](https://github.com/scikit-learn/scikit-learn/tree/master/sklearn/neighbors)

__Clustering__
* [K-Mean Clustering](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/k_means_.py)
* [Mean-Shift Clustering](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/mean_shift_.py)
* [Spectral Clustering](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/spectral.py)
* [Agglomerative Clustering](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/hierarchical.py)

Here are some methods found in Visumap that aren't in either of these packages, but would be really cool to have
* [Curvilinear Component Analysis](http://www.sciencedirect.com/science/article/pii/S0020025512006408)
* [Stochastic Neighbor Embedding](http://lvdmaaten.github.io/tsne/)
* [Self-Organizing Map](http://en.wikipedia.org/wiki/Self-organizing_map)
* [Self-Organzing Graph](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.37.6856&rep=rep1&type=pdf)

Also may want to consider the following non-linear ordination methods found in scikit bio
* [kernal PCA](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/decomposition/kernel_pca.py)
* [isomap](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/manifold/isomap.py)
* [probabilitistic PCA](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/decomposition/factor_analysis.py)

But before we go about turning this into a TODO list, I think we should consider the following

1. Where would be preprocessing procedures live?  Should they be placed in QIIME2?
2. If the above answer is yes, then we will need to redesign the current eigenvector file format to potentially hold 2 sets of eigenvectors rather than just 1

**Comment by Jorge-C** [comment]
Maybe the computations should be done in skbio?

**Comment by ElDeveloper** [comment]
A few of these look really interesting, thanks for condensing this into
an issue :+1:. The way I see it is that clustering and classification
could be supplementary to the ordination methods. For example k-means on
its own is entirely irrelevant, unless it is put in the context of a
visualization, statistical test or some other way of interpreting the
information. The same is true for classification, it doesn't make much
sense on its own either.

As for the file formats, I believe all the file formats should be HDF5
based, this would allow for a more flexible structure than what we have
through text files. But this is something that needs to be discussed
with the rest of the skbio devs.

Now that we are talking about other cool methods, what I would really
like to see is convex hulls and minimum spanning ellipsoids, see #18.

Regardless of the methods that we decide to include, the bigger
conversation is how are we going to be moving forward, the python
objects will live in skbio (no questions asked), however who provides
the CLI (if there's any) is the 1MD question. For the sake of simplicity
I would say QIIME2 does this, but it's tough to say this won't change in
the near future.

**Comment by ElDeveloper** [comment]
:+1:

**Comment by mortonjt** [comment]
Added the minimum spanning ellipsoid and convex hull visualization to the list above.

I completely agree with making classification/clustering supplementary to ordination.
If you take a look at the [3:00 mark in this Visumap demo](https://www.youtube.com/watch?v=pE5KUzHFSnk), you can see that there are different shapes representing different groups.
Maybe we can have a similar scheme for visualizing classification/clustering

As far as the file format goes, I think it would ideal to have the following features
* Eigenvectors for samples
* Eigenvectors for OTUs (optional)
* Eigenvalues

Here are a few of ideas to spur discussion
* Would we be interested in including sample/observation metadata in this file?  That way, we don't have to do the map file preprocessing in Emperor
* Would we want to specify sample/observation metadata colors/shapes/sizes in this file?
* When jack-knifing/bootstrapping, would we want to have all of the these eigenvectors/eigenvalues to be in a single file, or in a whole bunch of files (like it is now)?

I've just been thinking, if we can resort to having a comprehensive file format that can have all of the plotting data, metadata and visualization features, that would definitely simplify the user interface for Emperor - all you would need to specify is which program you want to run and the input file.

**Comment by antgonza** [comment]
Note that http://biom-format.org/documentation/adding_metadata.html
already support sample and observation metadata. However, we are not
taking advantage of this feature anywhere else.

**Comment by mortonjt** [comment]
Yup.  That is a definite possibility - we could potentially use BIOM tables to store eigenvectors and corresponding metadata.  Definitely would be a path of less resistance

**Comment by ElDeveloper** [comment]
> I completely agree with making classification/clustering supplementary to ordination.
If you take a look at the 3:00 mark in this Visumap demo, you can see that there are different shapes representing different groups. Maybe we can have a similar scheme for visualizing classification/clustering

Yes this would be awesome! see #128.

> Would we be interested in including sample/observation metadata in this file? That way, we don't have to do the map file preprocessing in Emperor
That makes perfect sense, though not a lot of people do this through BIOM, except for [Phinch](http://phinch.org).

> Would we want to specify sample/observation metadata colors/shapes/sizes in this file?

It would definitely be helpful! But not a requirement.

> When jack-knifing/bootstrapping, would we want to have all of the these eigenvectors/eigenvalues to be in a single file, or in a whole bunch of files (like it is now)?

I would oppose to having them in separate files, if we keep the files very simple, I think it will be better on the long run.

**Comment by Jorge-C** [comment]
I have no idea if this is feasible, but a cool option could be to have emperor consume python objects somehow and avoid text files altogether.

**Comment by ElDeveloper** [comment]
Hmmm I think that can be done with the emperor object, although not a
lot of the features exposed through the command line are available, the
bare-bones implementation is available here:
http://biocore.github.io/emperor/build/html/generated/emperor.core.Emperor.html#emperor.core.Emperor

```
from emperor import Emperor
# hack the world
```

The idea of this was to be able to have an IPython html repr (which is
implemented and works), but then this happened:
https://github.com/jupyter/nbviewer/issues/316

**Comment by Jorge-C** [comment]
Too bad... We should check again with the recently-released 3.0.0 IPython
notebook.

2015-03-06 14:48 GMT-07:00 Yoshiki Vázquez Baeza <notifications@github.com>:

**Comment by mortonjt** [comment]
I raised a new issue on how to handle visualization options here: https://github.com/biocore/emperor/issues/357

I've also thought about how to handle jack-knifing/bootstrapping - could we require the user to put all of the resampled samples into a single table?  This would allow us to ditch the option to input a directory for input_coords

**Comment by ElDeveloper** [comment]
I just tried out with the latest release version of the ipython notebook
and it didn't work. :sad:

**Comment by Jorge-C** [comment]
:cry:

The problem only happens when sharing a notebook via nbviewer/nbconvert right? I pinged the jupyter issue just in case.

**Comment by ElDeveloper** [comment]
Yeah, it still is, thanks for doing that!

Yoshiki Vázquez-Baeza

**Comment by mortonjt** [comment]
Here's another thought,

Would it be a good idea to have a publicly accessible preprocessing API in emperor?
I'm thinking it would have the following benefits

* Could potentially alleviate complete dependence on skbio/QIIME.  We could had preprocessing functions that can accept any set of eigenvector/eigenvalues to plot, which can make it easier for users to use other dimensionality reduction methods (e.g. methods from scikit-learn)
* Will prevent bloating in skbio.  I think if we start porting over all of our helper functions into skbio (e.g. biplot calculation), this will bloat the skbio API, making it harder to maintain.

**Comment by ElDeveloper** [comment]
Do you mean something like the Emperor object in core.py? Admittedly
that class needs to expand a bit more to cover the full interface
provided through the CLI.

**Comment by mortonjt** [comment]
Exactly

**Comment by ElDeveloper** [comment]
:+1:, this would also help once we start moving to click for the
creation of the command line interface.

**Comment by ConstantinoSchillebeeckx** [comment]
+1 for methods for clustering.  Currently I'm doing a pre-processing step on my mapping file using [sklearn.cluster.kmeans](http://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html) to generate k clusters.
![screen shot 2016-02-24 at 11 01 27](https://cloud.githubusercontent.com/assets/8518288/13293655/200e3b84-dae6-11e5-8793-3cee5ce4f262.png)
